base/lock.jl

```diff
@@ -983,3 +983,130 @@ end
+
+The depot is determined by the first entry in `DEPOT_PATH` that exists and is writable.
+Token files are stored in `depot/tokens/` and use `mkpidlock` to ensure atomic initialization
+across processes.
+
+# Arguments
+- `init::Function`: The function to call to initialize the value.
+- `token_name::AbstractString`: The name for the token file.
+
+!!! note "Representation limitations"
+    Values are stored using `repr()` and restored using `Meta.parse()`. This works well for
+    basic Julia types (numbers, strings, arrays, etc.) but may not work for complex objects
+    that don't have a parseable string representation.
+
+!!! compat "Julia 1.13"
+    This type requires Julia 1.13 or later.
+
+## Example
+
+```jldoctest
+julia> const depot_cache = Base.OncePerDepot{Vector{String}}("mycache") do
+           println("Initializing cache...")
+           return ["depot-specific", "data"]
+       end;
+
+julia> depot_cache() |> typeof
+Initializing cache...
+Vector{String} (alias for Array{String, 1})
+
+julia> depot_cache() === depot_cache()  # Same within depot
+true
+```
+"""
+mutable struct OncePerDepot{T, F} <: Function
+    const initializer::F
+    const token_name::String
+
+    OncePerDepot{T}(initializer::Type{U}, token_name::AbstractString) where {T, U} = new{T,Type{U}}(initializer, String(token_name))
+    OncePerDepot{T}(initializer::F, token_name::AbstractString) where {T, F} = new{T,F}(initializer, String(token_name))
+    OncePerDepot{T,F}(initializer::F, token_name::AbstractString) where {T, F} = new{T,F}(initializer, String(token_name))
+    OncePerDepot(initializer::Type{U}, token_name::AbstractString) where U = new{Base.promote_op(initializer), Type{U}}(initializer, String(token_name))
+    OncePerDepot(initializer, token_name::AbstractString) = new{Base.promote_op(initializer), typeof(initializer)}(initializer, String(token_name))
+end
+
+function (once::OncePerDepot{T,F})() where {T,F}
+    if !@isdefined(mkpidlock_hook)
+        error("mkpidlock not available, OncePerDepot is not safe across processes")
+    end
+    # Find the first writable depot
+    depot = nothing
+    for path in DEPOT_PATH
+        if isdir(path) && Sys.iswritable(path)
+            depot = path
+            break
+        end
+    end
+
+    if depot === nothing
+        error("No writable depot found in DEPOT_PATH")
+    end
+
+    # Create tokens directory if it doesn't exist
+    tokens_dir = joinpath(depot, "tokens")
+    if !isdir(tokens_dir)
+        try
+            mkpath(tokens_dir)
+        catch
+            if !isdir(tokens_dir)  # Race condition check
+                rethrow()
+            end
+        end
+    end
+
+    token_path = joinpath(tokens_dir, once.token_name)
+    value_path = token_path * ".value"
+
+    # Try to load existing value first
+    if isfile(value_path)
+        try
+            value_str = read(value_path, String)
+            return Core.eval(Main, Meta.parse(value_str))::T
```

Comment:
Eeek, this runs arbitrary code from the file system in ways that are likely unexpected - I don't think this is a good idea.

Reply:
Good point. I wanted to use serialize/deserialize but that's not in base. This could be moved to a stdlib.

Reply:
Well, but it doesn't support arbitrary values already, since it's cross process - I think that's just a fundamental restriction to this kind of API. At the very least we need to have a strong handle on what the security implications of this are. It is true that we load arbitrary code from depots anyway, so maybe it's not that big of a deal, but I can just see this immediately becoming an exploitation vector if this directory is not sufficiently protected. At the very least needs to set a umask, etc, but if we can avoid the footgun entirely, I'd be strongly in favor.

Reply:
I think these tokens will need to be name-space-prefixed for the module they're defined in.. I don't currently have a good solution for that, other than making this a macro..